### PR TITLE
Solve task 11 FEATURE: GET /api/articles (queries)

### PR DESCRIPTION
**Pull request by isabelpopa**

### CODE REVIEW

File: __tests__/api/api.test.js [hunk 1]
```diff
@@ -165,6 +165,79 @@ describe("GET /api/articles", () => {
       });
   });
 });
+describe("FEATURE: GET /api/articles (queries)", () => {
+  test("Articles should be filtered by topic=mitch, sorted by author and ordered ascending ", () => {
+    return request(app)
+      .get("/api/articles?topic=mitch&sort_by=author&order=asc")
+      .then(({ body }) => {
+        const { articles } = body;
+        expect(articles).toBeSortedBy("author", { descending: false });
```
Comment (B7Bprog) [inline __tests__/api/api.test.js:176]: Please check the length of the array.

File: __tests__/api/api.test.js [hunk 2]
```diff
@@ -165,6 +165,79 @@ describe("GET /api/articles", () => {
...
+              "https://images.pexels.com/photos/158651/news-newsletter-newspaper-information-158651.jpeg?w=700&h=700",
+        })
+      });
+  });
+  test("Articles should be filtered by topic=cats, sorted by votes and ordered descending by default ", () => {
+    return request(app)
+      .get("/api/articles?topic=cats&sort_by=votes")
+      .then(({ body }) => {
+        const { articles } = body;
+        expect(articles).toBeSortedBy("votes", { descending: true });
```
Comment (B7Bprog) [inline __tests__/api/api.test.js:209]: Length of the array needs to be checked here as well.

File: __tests__/api/api.test.js [hunk 3]
```diff
@@ -165,6 +165,79 @@ describe("GET /api/articles", () => {
...
+  test("Articles should be filtered by topic=mitch, sorted by author and ordered ascending ", () => {
+    return request(app)
+      .get("/api/articles?topic=mitch&sort_by=author&order=asc")
+      .then(({ body }) => {
+        const { articles } = body;
+        expect(articles).toBeSortedBy("author", { descending: false });
+        articles.forEach((article) => {
+          expect(article).toHaveProperty("topic", "mitch");
+        });
+        expect(articles[0]).toStrictEqual({
```
Comment (B7Bprog) [inline __tests__/api/api.test.js:189]: You could check all articles here, also using a more flexible matcher like .toMatchObject() or .objectContaining() to allow future expansion of articles.

File: __tests__/api/api.test.js [hunk 4]
```diff
@@ -165,6 +165,79 @@ describe("GET /api/articles", () => {
...
+  test("400: responds with 400 and 'Bad sort_by request'", () => {
+    return request(app)
+      .get("/api/articles?sort_by=banana")
+      .expect(400)
+      .then(({ body }) => {
+        const { msg } = body;
+        expect(msg).toBe("Bad sort_by request");
+      });
+  });
+  test("400: responds with 400 and 'Bad topic request'", () => {
```
Comment (B7Bprog) [inline __tests__/api/api.test.js]: This should be a 404 as a new topic called 'banana' could be added later.

File: __tests__/api/api.test.js [hunk 5]
```diff
@@ -165,6 +165,79 @@ describe("GET /api/articles", () => {
...
+    return request(app)
+      .get("/api/articles?topic=banana")
+      .expect(400)
+      .then(({ body }) => {
+        const { msg } = body;
+        expect(msg).toBe("Bad topic request");
+      });
+  });
+  test("404: responds with 404 and 'Not Found' when topic is valid but not found", () => {
```
Comment (B7Bprog) [inline __tests__/api/api.test.js]: This test should pass even the 'coding' is not in your 'acceptedValuesTopic' array. To check what topics you already have, you could query the database in the model for that.

File: __tests__/api/api.test.js [hunk 6]
```diff
@@ -165,6 +165,79 @@ describe("GET /api/articles", () => {
       });
   });
 });
+describe("FEATURE: GET /api/articles (queries)", () => {
+  test("Articles should be filtered by topic=mitch, sorted by author and ordered ascending ", () => {
+    return request(app)
+      .get("/api/articles?topic=mitch&sort_by=author&order=asc")
```
Comment (B7Bprog) [inline __tests__/api/api.test.js:171]: Please check the status code here.

File: __tests__/api/api.test.js [hunk 7]
```diff
@@ -165,6 +165,79 @@ describe("GET /api/articles", () => {
...
+            author: "butter_bridge",
+            created_at: "2020-07-09T20:11:00.000Z",
+            article_img_url:
+              "https://images.pexels.com/photos/158651/news-newsletter-newspaper-information-158651.jpeg?w=700&h=700",
+        })
+      });
+  });
+  test("Articles should be filtered by topic=cats, sorted by votes and ordered descending by default ", () => {
+    return request(app)
+      .get("/api/articles?topic=cats&sort_by=votes")
```
Comment (B7Bprog) [inline __tests__/api/api.test.js:204]: Check status code here as well.

File: __tests__/api/api.test.js [hunk 8]
```diff
@@ -165,6 +165,79 @@ describe("GET /api/articles", () => {
...
+          comment_count: "2",
+          title: "UNCOVERED: catspiracy to bring down democracy",
+          topic: "cats",
+          author: "rogersop",
+          created_at: "2020-08-03T13:14:00.000Z",
+          article_img_url:
+            "https://images.pexels.com/photos/158651/news-newsletter-newspaper-information-158651.jpeg?w=700&h=700",
+        })
+      });
+  });
```
Comment (B7Bprog) [inline __tests__/api/api.test.js:225]: Please write another test for a successful response where there is a topic but it doesn't have any articles.

File: __tests__/api/api.test.js [hunk 9]
```diff
@@ -165,6 +165,79 @@ describe("GET /api/articles", () => {
...
+  test("400: responds with 400 and 'Bad sort_by request'", () => {
+    return request(app)
+      .get("/api/articles?sort_by=banana")
+      .expect(400)
+      .then(({ body }) => {
+        const { msg } = body;
+        expect(msg).toBe("Bad sort_by request");
+      });
+  });
```
Comment (B7Bprog) [inline __tests__/api/api.test.js:244]: Please add a test, where the order is invalid. For example: ?order=bananas

**Review by B7Bprog — CHANGES_REQUESTED** [review]
Hey Isabel, please see my comments, refactor / add tests, then please send us another PR with it.

**Review by Hard-Luck — APPROVED** [review]
Great work, taking the feedback on board.

**Pull request merged**